Enforce RFC5280 extension rules in CRL app and check for duplicate X509v3 extensions

Th3BanHamm3r opened the pull request:
crl_verify: enforce RFC5280 extension rules in do_ver()
- Count and tabulate all CRL extension NIDs present
- Require exactly one instance of Authority Key Identifier and CRL Number
- Allow at most one instance of IssuerAltName, DeltaCRLIndicator,
  IssuingDistributionPoint, FreshestCRL, and AuthorityInfoAccess
- Reject the CRL if any unknown extension is marked critical
- Return errors and abort verification on violations

@bbbrumley @t8m @owen5280

Fixes #26661, #25799

##### Checklist
- [x] documentation is added or updated
- [ ] tests are added or updated

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -231,7 +236,109 @@ int crl_main(int argc, char **argv)
             goto end;
         }
-
+        /* Count CRL extensions */
```
With this addition in mind, would it be better to leave it in the verify check flag or add another flag dedicated to extension verification? If so, this feature could also address https://github.com/openssl/openssl/issues/25799, https://github.com/openssl/openssl/issues/27114, https://github.com/openssl/openssl/issues/27160, https://github.com/openssl/openssl/issues/27321, https://github.com/openssl/openssl/issues/27360, and https://github.com/openssl/openssl/issues/27474. That being said, it will only do so in this CRL app, not when a CRL is actually being used. From prior comments in the above issues, it seems like OpenSSL's stance is that it's up to the CA to create standard CRLs and that they should be parsed normally. And if the user wants to verify a particular CRL, they can do so within the CRL app. @bbbrumley @t8m

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -338,6 +338,47 @@ int crl_main(int argc, char **argv)
+                goto end;
+            }
+            if (akid->serial != NULL
+                && ASN1_INTEGER_get(akid->serial) <= 0) {
```
Addresses #27321

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -338,6 +338,47 @@ int crl_main(int argc, char **argv)
+                ret = 1;
+                goto end;
+            }
+            if (akid->keyid->length == 0) {
```
Addresses #27474, #27360

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -338,6 +338,47 @@ int crl_main(int argc, char **argv)
+        akid = X509_CRL_get_ext_d2i(x, NID_authority_key_identifier,
+                                    &crit, &idx);
+        if (akid) {
+            if (crit) {
```
Addresses #27160

t8m commented:
As this is a new functionality we would like to see some testcases for it. Would you please add them to 25-test_crl.t ?

Th3BanHamm3r commented:
> As this is a new functionality we would like to see some testcases for it. Would you please add them to 25-test_crl.t ?

Yes, of course. I wanted to ensure I checked with the maintainers that this is the implementation you all were looking for before I added test cases and whatnot. There was also mention of adding this to X509_V_FLAG_X509_STRICT, so just double-checking that this was the intended implementation. @t8m @bbbrumley

vdukhovni commented on apps/crl.c:
```diff
@@ -231,7 +236,150 @@ int crl_main(int argc, char **argv)
+        /* Additional Authority Key Identifier semantic checks */
+        crit = 0;
+        idx = -1;
+        akid = X509_CRL_get_ext_d2i(x, NID_authority_key_identifier,
+                                    &crit, &idx);
+        if (akid) {
+            if (crit) {
+                BIO_printf(bio_err,
+                           "CRL extension Authority Key Identifier "
+                           "must not be marked critical\n");
+                AUTHORITY_KEYID_free(akid);
+                ret = 1;
+                goto end;
+            }
```
I am opposed to this change. It serves no security purpose.  This is mere pedantic enforcement of interoperability requirements on issuing CAs that should not be enforced by CRL consumers.

vdukhovni commented on apps/crl.c:
```diff
@@ -231,7 +236,150 @@ int crl_main(int argc, char **argv)
+                case NID_freshest_crl:
+                case NID_info_access:
+                    if (cnt > 1) {
+                        BIO_printf(bio_err,
+                                   "CRL extension %s must appear "
+                                   "at most once (found %d)\n",
+                                   sn, cnt);
+                        ret = 1;
+                        OPENSSL_free(counts);
```
This should be a warning, not a failure to parse the CRL, perhaps with a flag that makes the warning fatal?. What is the purpose of refusing the parse the CRL?
Indeed extensions MUST not appear more than once in a certificate or CRL, but the `crl(1)` utility should not be where this is enforced. 
When a CRL has problems, a user should be able to inspect it with this utility, and failing to parse is likely counterproductive.

The place to reject ambiguous CRLs (as a result of multiple instances of an extension) is when the CRL is used as part of certificate validation. In that context we need to exercise judgment about which issues are relevant to enforce when verifying rather than advice to CAs to ensure interoperability.

vdukhovni commented on apps/crl.c:
```diff
@@ -338,6 +338,47 @@ int crl_main(int argc, char **argv)
+                ret = 1;
+                goto end;
+            }
+            if (akid->keyid == NULL) {
```
This restriction should not be enforced when consuming CRLs.

bbbrumley commented:
@Th3BanHamm3r you're gonna want to rebase to remove the merge commit (per project guidelines). Cheers

bbbrumley commented:
... and @vdukhovni @t8m tyvm for the feedback :pray:

vdukhovni commented:
> ... and @vdukhovni @t8m tyvm for the feedback 🙏

You're welcome.  Note that I'm generally quite sceptical of interpreting RFC5280 (inteoperability) constraints on what a conformant CA can sign as security-relevant requirements for verifiers to check.

Barring a compelling reason, OpenSSL should not be needlessly strict here.

I do agree that duplicate occurrence of a given extension makes a certificate or CRL ambiguous, which is security-relevant, so that should be enforced when verifying, but likely not when parsing, printing, ... So any changes that are warranted are more likely in crypto/x509/x509_vfy.c than in `apps/*.c`.

Th3BanHamm3r committed:
crl/x509: enforce full RFC5280 extension rules and comprehensive CRL tests

- apps/crl: add `-check-extensions` flag to emit diagnostics for all CRL extension rules and warning output.
  - Authority Key Identifier (cardinality & criticality, empty keyIdentifier, authorityCertIssuer, authorityCertSerialNumber)
  - CRL Number (cardinality & criticality)
  - Issuer Alternative Name (duplicates & criticality)
  - Delta CRL Indicator (criticality & duplicates)
  - Issuing Distribution Point (criticality & duplicates)
  - Freshest CRL (invalid in delta, duplicates & criticality)
  - Authority Information Access (duplicates & criticality)
  - unknown critical extensions
- x509_vfy: add duplicate-OID rejection in certs and CRLs when `X509_V_FLAG_X509_STRICT` is enabled.
  - define `X509_V_ERR_DUPLICATE_EXTENSION` (96) in `x509_vfy.h.in`
  - add error message in `x509_txt.c`
  - handle case in `verify.c`
- 25-test_crl: add RFC5280 CRL extension enforcement tests
- 25-test_verify: add duplicate extension enforcement tests

Th3BanHamm3r commented:
@vdukhovni @t8m @bbbrumley
I have updated the CRL app to include all 5280 rules under a new '-extension' flag. That way, extensions can be verified separately. I'd like to get a sanity check on them as well to make sure the logic is clean, if possible. Assuming that this is acceptable, I will then go in and add tests accordingly in 25-test_crl.t and update the documentation.
As for adding checks to crypto/x509/x509_vfy.c, should this be done in another PR?

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -231,7 +236,150 @@ int crl_main(int argc, char **argv)
+                name = OBJ_nid2sn(counts[i].nid);
+                if (name == NULL)
+                    name = OBJ_nid2ln(counts[i].nid);
+                BIO_printf(bio_err,
+                           "Extension %-30s (NID %3d): %d\n",
+                           name,
+                           counts[i].nid,
+                           counts[i].count);
```
Sounds good, I will leave as is under the now changed '-enforce_exts' flag.

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -250,11 +260,224 @@ int crl_main(int argc, char **argv)
+            }
+            for (i = 0; i < ext_total; i++) {
+                ext = sk_X509_EXTENSION_value(exts, i);
+                nid = OBJ_obj2nid(X509_EXTENSION_get_object(ext));
```
Output should now resolve to:
Extension X509v3 CRL Number                   (OID 2.5.29.20): 1
Extension X509v3 Issuing Distribution Point   (OID 2.5.29.28): 1
Extension X509v3 Authority Key Identifier     (OID 2.5.29.35): 2
Extension X509v3 Issuer Alternative Name      (OID 2.5.29.18): 1
Extension X509v3 Delta CRL Indicator          (OID 2.5.29.27): 1
Extension X509v3 Freshest CRL                 (OID 2.5.29.46): 1
Extension Authority Information Access        (OID 1.3.6.1.5.5.7.1.1): 1

Except it's prettier than the alignment here. lol

Th3BanHamm3r commented:
The check for OID duplication was added to x509_vfy.c and supporting files for both CRLs (check_crl) and certificates (check_extensions). They utilize the same static function to do so. @vdukhovni, checking to see if this is an ok implementation.

As for adding tests, @t8m are you looking for a test that checks every possible 5280 missconfiguration? 
I assume a test is needed for the duplication checks x509_vfy.c does as well. Does adding the one with duplicate AKI, as mentioned in #26661, suffice?

The CRL provided in that issue is also interesting when run through CRL with -enforce_ext. It flags a whole bunch of things:
! Extension Authority Key Identifier (id-ce 35) MUST appear exactly once (found 2)
! Extension Issuing Distribution Point MUST be marked critical
! Extension Authority Key Identifier MUST NOT include authorityCertIssuer when using key-ID
! Extension Authority Key Identifier MUST NOT include authorityCertSerialNumber when using key-ID
! Extension Authority Key Identifier MUST NOT be marked critical
! Extension Delta CRL Indicator MUST be marked critical in a delta CRL
! Extension Freshest CRL MUST NOT appear in a delta CRL

Do the tests need to map 1 to 1, or can this one CRL be used to test all 7 of these (as an example)?

t8m commented:
For the test to be useful there has to be a separate CRL for every newly added CRL check. So 7 CRLs with 7 different issues.

Th3BanHamm3r commented:
@t8m I've added the tests to the 25-text_crl.t file. I'm not super familiar with perl so if you could take a look over to make sure im not messing anything up I'd appreciate it.
As for adding the .pem files themselves, do you know of an easy way to create these broken CRLs? The openssl tool doesn't allow for certain fields to be misconfigured.

bbbrumley commented:
> do you know of an easy way to create these broken CRLs

How many can you lift from the initial issue reports on GH?

I would personally use a combination of `openssl asn1parse -genconf` and the awesome community tool [asn1template](https://github.com/wllm-rbnt/asn1template) to do this. I admit, this requires a certain level of masochism :shrug:

Th3BanHamm3r commented:
> How many can you lift from the initial issue reports on GH?
From the reports, I might be able to lift most of them. I'd have to check that .der is okay to use, though (the existing tests use .pem). To do that, though, I would have to figure out how to split them apart since most of the certs given contain more than one error, and I'm not sure that's possible. It might just be easier to make each one individually. Thanks for the insight @bbbrumley, I'll have to take a look.

As for the duplicate test required for the x509_vfy.c change, checking that the correct test file to add to is 25-test_x509.t?

t8m commented on apps/crl.c:
```diff
@@ -59,6 +59,7 @@ const OPTIONS crl_options[] = {
     {"crlnumber", OPT_CRLNUMBER, '-', "Print CRL number"},
     {"badsig", OPT_BADSIG, '-', "Corrupt last byte of loaded CRL signature (for test)" },
     {"gendelta", OPT_GENDELTA, '<', "Other CRL to compare/diff to the Input one"},
+    {"enforce_exts", OPT_ENFORCE_EXTS, '-',  "Perform RFC-5280 CRL extension checks"},
```
IMO something like `check_extensions` would be a better option name.

t8m commented:
> As for the duplicate test required for the x509_vfy.c change, checking that the correct test file to add to is 25-test_x509.t?

IMO 25-test_verify.t would be a better place.

shahsb reviewed (commented):
Incorrect Handling of Authority Key Identifier (AKID)

Problem: The approach requires exactly one AKID, but RFC5280 states AKID is optional (Section 5.2.1).

Fix: Remove the requirement for AKID unless explicitly needed for your use case (e.g., to match issuer keys in multi-key CAs). RFC5280 does not mandate AKID in CRLs.

### EVIDENCES: https://www.rfc-editor.org/rfc/rfc5280#section-5.2.1

[5.2.1](https://www.rfc-editor.org/rfc/rfc5280#section-5.2.1).  Authority Key Identifier

   The authority key identifier extension provides a means of
   identifying the public key corresponding to the private key used to
   sign a CRL.  The identification can be based on either the key
   identifier (the subject key identifier in the CRL signer's
   certificate) or the issuer name and serial number.  This extension is
   especially useful where an issuer has more than one signing key,
   either due to multiple concurrent key pairs or due to changeover.

shahsb reviewed (commented):
AuthorityInfoAccess (AIA) in CRLs

Problem: `AuthorityInfoAccess` is not a standard CRL extension. It is defined for certificates (RFC5280, Section 4.2.2.1). Including it here is likely a mistake.

Fix: Remove AuthorityInfoAccess from the allowed list. If referring to CRL distribution points, use CRLDistributionPoints (a certificate extension, not a CRL extension).

#### EVIDENCES: https://www.rfc-editor.org/rfc/rfc5280#section-4.2.2.1

[4.2.2.1](https://www.rfc-editor.org/rfc/rfc5280#section-4.2.2.1).  Authority Information Access

   The authority information access extension indicates how to access
   information and services for the issuer of the certificate in which
   the extension appears.  Information and services may include on-line
   validation services and CA policy data.  (The location of CRLs is not
   specified in this extension; that information is provided by the
   cRLDistributionPoints extension.)  This extension may be included in
   end entity or CA certificates.  Conforming CAs MUST mark this
   extension as non-critical.

shahsb reviewed (commented):
`IssuingDistributionPoint` MUST be critical if present (RFC5280, Section 5.2.5). The approach does not verify this.

`CRLNumber` and `FreshestCRL` MUST NOT be critical (RFC5280, Sections 5.2.3 and 5.2.6). The approach allows them to be critical.

Fix: Add checks for criticality of known extensions:

```
if IssuingDistributionPoint is present and not critical → reject
if CRLNumber is critical → reject
if FreshestCRL is present and critical → reject
```

vdukhovni commented on apps/crl.c:
```diff
@@ -250,11 +262,246 @@ int crl_main(int argc, char **argv)
+                    saw_delta = counts[i].count;
+            }
+            /* Ensure required extensions are present exactly once */
+            if (saw_aki != 1) {
```
Once again, I DO NOT support pedantic RFC checks that are not security-relevant. These are interoperability requirements on the issuer CA, but MUST NOT be enforced by OpenSSL.  Please do not add **any** such checks.

Th3BanHamm3r replied:
CRL number is reported as missing. AKI is not reported since a 0 instance means self-signed. Other checks are optionally enabled with the -check-extensions flag.

shahsb reviewed (commented):
**Problem:** DeltaCRLIndicator must be present exactly once in delta CRLs and absent in base CRLs (RFC5280, Section 5.2.4). The approach allows at most one instance but does not validate based on CRL type.

**Fix:** After counting extensions, check:

If DeltaCRLIndicator is present, ensure it is a delta CRL.

If DeltaCRLIndicator is absent, ensure it is a base CRL.

**Evidence/Proof:** https://www.rfc-editor.org/rfc/rfc5280#section-5.2.4

shahsb reviewed (commented):
Missing Extensions: Consider handling CRLReason (for revoked certificates) and CertificateIssuer (for indirect CRLs), though these are certificate entry extensions, not CRL extensions.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

bbbrumley commented:
The window to motivate OP to make changes to this PR is gone, so I suggest just closing this one out

Th3BanHamm3r commented:
> The window to motivate OP to make changes to this PR is gone, so I suggest just closing this one out

No worries, I'm still working on this, have just been busy with a few 6 week summer classes that have been eating my time. They finished this last Thursday so planning on reviewing the comments and pushing an update either Monday or Tuesday. 👍 Sorry for the wait.

DDvO reviewed (changes requested):
Thank you for taking this up.

I did not look into much detail,
but found various mostly high-level issues to be handled.

Please note also that there is a list of coding style nits reported by CI.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

Th3BanHamm3r commented on crypto/x509/x509_vfy.c:
```diff
@@ -570,13 +571,48 @@ static int check_purpose(X509_STORE_CTX *ctx, X509 *x, int purpose, int depth,
+ *  0 if a duplicate is found,
+ * −1 on out‐of‐memory.
+ */
+static int has_no_dup_ext_oids(const STACK_OF(X509_EXTENSION) *exts)
```
I saw the DOC changes last week. What still needs to be addressed?

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -250,11 +262,246 @@ int crl_main(int argc, char **argv)
+                    if (saw_delta > 0)
+                        BIO_printf(bio_err,
+                                   "! Extension Freshest CRL MUST NOT "
+                                   "appear in a delta CRL\n");
```
Showing when -check-extension flag enabled

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -250,11 +262,246 @@ int crl_main(int argc, char **argv)
+                    if (crit)
+                        BIO_printf(bio_err,
+                                   "! Extension %s SHOULD NOT be critical\n",
+                                   sn);
```
Showing when -check-extension flag enabled

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -250,11 +262,246 @@ int crl_main(int argc, char **argv)
+                    if (crit)
+                        BIO_printf(bio_err,
+                                   "! Extension CRL Number MUST NOT be "
+                                   "marked critical\n");
```
Showing when -check-extension flag enabled

Th3BanHamm3r commented on apps/crl.c:
```diff
@@ -250,11 +262,246 @@ int crl_main(int argc, char **argv)
+                if (ext0)
+                    crit = X509_EXTENSION_get_critical(ext0);
+                /*
+                * Enforce RFC 5280 §5.1.2.7 (Extensions field) and §5.2 (CRL Extensions) rules:
```
Correct, extension duplication is already dealt with in x509_vfy , checks have been removed. For the strictness, since this is an optional flag to check a CRL's compliance, I would argue they should be flagged when -check-extensions is enabled.

Th3BanHamm3r committed:
fixes: addressed comments and documentation

- crl.c: AKI is not required, duplicates are taken care of my x509_verify.c among other changes.
- -check-extensions flag added to docs
- duplicate extensions check added to x509_verify & verification options docs

Th3BanHamm3r commented:
@DDvO @vdukhovni @t8m I think I've addressed the noted problems. Though, I'm still struggling to figure out how to get specific CRLs for the tests.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 247 days ago

andrewkdinh commented:
hi, can you still work on this? @Th3BanHamm3r

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago
